### PR TITLE
connectivity: Introduce Multicast connectivity test

### PR DESCRIPTION
**Pull request by yushoyamaguchi**
Introduces Multicast connectivity test, working with netshoot container.
Using socat for IGMP and UDP communication test.
Receivers are running as Daemonset and a sender is running as a Deployment which has 1 replicaset.

Operation was confirmed on the kind and kubeadm clusters.

This if follow up of https://github.com/cilium/cilium-cli/issues/2615 .

In addtion, this PR is related to https://github.com/cilium/cilium-cli/pull/2620 .
I want to implement a test scenario using multicast subcommand discussing in above PR after this is merged.

I have the following questions and would like to discuss here and fix them.

<details>
<summary>related part of this command output (when 3 nodes)</summary>

```
[=] [cilium-test-1] Test [multicast] [100/103]
  [.] Action [multicast/multicast/socat multicast]
  🐛 Executing command [timeout 1 socat STDIO UDP4-RECVFROM:6666,ip-add-membership=239.255.9.9:10.244.0.10]
.  [.] Action [multicast/multicast/socat multicast]
  🐛 Executing command [timeout 1 socat STDIO UDP4-RECVFROM:6666,ip-add-membership=239.255.9.9:10.244.2.142]
.  [.] Action [multicast/multicast/socat multicast]
  🐛 Executing command [timeout 1 socat STDIO UDP4-RECVFROM:6666,ip-add-membership=239.255.9.9:10.244.1.88]
.  🐛 Finalizing Test multicast
  🐛 Finalizing Test host-firewall-ingress
  🐛 Finalizing Test host-firewall-egress
  🐛 Finalizing Test check-log-errors

✅ [cilium-test-1] All 1 tests (3 actions) successful, 102 tests skipped, 0 scenarios skipped.
```

</details>

### CODE REVIEW

**Comment by yushoyamaguchi** [comment]
image

This is the output of the test with `--debug`.

After rebasing the code, 
`❌ writing to junit file  failed: open : no such file or directory`
This error is occurred.

Maybe, the cause is this commit. (https://github.com/cilium/cilium-cli/commit/fa5be3a0f5563ae6f82bf47ff4d8c7339c5a4940)
The way of testing is changed or some additional setting is required?

File: connectivity/check/deployment.go [hunk 1]
```diff
@@ -1417,6 +1449,38 @@ func (ct *ConnectivityTest) validateDeployment(ctx context.Context) error {
 		}
 	}
 
+	if ct.Features[features.Multicast].Enabled {
+		// socat client pods
+		/*if err := WaitForDeployment(ctx, ct, ct.clients.src, ct.Params().TestNamespace, netshootSocatClientDeploymentName); err != nil {
```
Comment (tommyp1ckles) [inline connectivity/check/deployment.go:1454]: Unused code?

**Review by tommyp1ckles — CHANGES_REQUESTED** [review]

**Comment by yushoyamaguchi** [comment]
In addition, maybe we have to think about it. (https://github.com/cilium/cilium-cli/pull/2724)
This multicast connectivity test is mutate cilium eBPF map temporary. (After testing, it's going back to the original state.)

**Comment by tommyp1ckles** [comment]
I'll do a general code review, however multicast isn't really my area of expertise. @ldelossa Could you take a look as well?

File: cli/connectivity.go [hunk 2]
```diff
@@ -151,6 +151,7 @@ func newCmdConnectivityTest(hooks api.Hooks) *cobra.Command {
 	cmd.Flags().StringVar(&params.DNSTestServerImage, "dns-test-server-image", defaults.ConnectivityDNSTestServerImage, "Image path to use for CoreDNS")
 	cmd.Flags().StringVar(&params.TestConnDisruptImage, "test-conn-disrupt-image", defaults.ConnectivityTestConnDisruptImage, "Image path to use for connection disruption tests")
 	cmd.Flags().StringVar(&params.FRRImage, "frr-image", defaults.ConnectivityTestFRRImage, "Image path to use for FRR")
+	cmd.Flags().StringVar(&params.NetshootImage, "netshoot-image", defaults.ConnectivityTestNetshootImage, "Image path to use for netshoot")
```
Comment (tommyp1ckles) [inline cli/connectivity.go:154]: I know we haven't really been doing this, but could you expand the flag description to mention what this image is used for?

File: connectivity/check/deployment.go [hunk 3]
```diff
@@ -1042,6 +1042,34 @@ func (ct *ConnectivityTest) deploy(ctx context.Context) error {
 		}
 	}
 
+	if ct.Features[features.Multicast].Enabled {
+		_, err = ct.clients.src.GetDeployment(ctx, ct.params.TestNamespace, netshootSocatClientDeploymentName, metav1.GetOptions{})
+		if err != nil {
```
Comment (tommyp1ckles) [inline connectivity/check/deployment.go:1047]: Should we check if this is not ErrNotExist?

File: connectivity/check/deployment.go [hunk 4]
```diff
@@ -1200,6 +1228,8 @@ func (ct *ConnectivityTest) deploymentList() (srcList []string, dstList []string
 		srcList = append(srcList, lrpBackendDeploymentName)
 	}
 
+	// Q: How about multicast deployments?
```
Comment (tommyp1ckles) [inline connectivity/check/deployment.go:1231]: Is this leftover?

File: connectivity/tests/multicast.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+			}
+		}(ciliumPod)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	var errRet error
+	for fetchdata := range errCh {
+		if fetchdata != nil {
```
Comment (tommyp1ckles) [inline connectivity/tests/multicast.go:304]: [errors.Join discards nil values](https://pkg.go.dev/errors#Join), so you likely don't need this nil check here.

File: connectivity/tests/multicast.go [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+	}
+	ciliumPods := ciliumPodsList.Items
+
+	for _, ciliumPod := range ciliumPods {
+		if nodeName == ciliumPod.Spec.NodeName {
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "group", "delete", testMulticastGroupIP}
+			_, stdErr, err := client.ExecInPodWithStderr(ctx, ciliumPod.Namespace, ciliumPod.Name, defaults.AgentContainerName, cmd)
+			if err != nil {
+				if !strings.Contains(stdErr.String(), notExistMsg) {
+					errMsg := fmt.Sprintf("Error: %v, Stderr: %s", err, stdErr.String())
```
Comment (tommyp1ckles) [inline connectivity/tests/multicast.go:328]: minor thing, can we include the testMulticastGroupIP here, might make it easier for someone to debug issues in the future.

File: connectivity/tests/multicast.go [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+	}
+	ciliumPods := ciliumPodsList.Items
+
+	for _, ciliumPod := range ciliumPods {
+		if nodeName == ciliumPod.Spec.NodeName {
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "subscriber", "delete", testMulticastGroupIP, subscriberIP}
+			_, stdErr, err := client.ExecInPodWithStderr(ctx, ciliumPod.Namespace, ciliumPod.Name, defaults.AgentContainerName, cmd)
+			if err != nil {
+				if !strings.Contains(stdErr.String(), notExistMsg) {
+					errMsg := fmt.Sprintf("Error: %v, Stderr: %s", err, stdErr.String())
```
Comment (tommyp1ckles) [inline connectivity/tests/multicast.go:355]: Same as above, but with mcast ip and subscriber ip.

**Review by tommyp1ckles — COMMENTED** [review]
Changes make sense to me, but I'll defer to someone who knows a bit more about Cilium multicast.

**Comment by yushoyamaguchi** [comment]
@tommyp1ckles 
Thank you for your review.
I've fixed many places which I can resolve.
I still have some questions for fix, so I'm waiting opinions of multicast expert.

P.S.
Have this discussion finished? : https://github.com/cilium/cilium-cli/pull/2726
I think whether this PR is acceptable or not depends on the outcome of that discussion.

**Pull request closed without merging**